pitop/robotics/navigation/navigation_controller.py

```diff
@@ -0,0 +1,407 @@
+import time
+import numpy as np
+import math
+from typing import Union
+from simple_pid import PID
+import sched
+from inspect import getfullargspec
+
+
+class RobotState:
+    def __init__(self):
+        self._x = 0.0      # m
+        self._y = 0.0      # m
+        self._angle = 0.0  # degrees
+        self._theta = 0.0  # radians
+        self.v = 0.0       # m/s
+        self.w = 0.0       # rad/s
+
+    def __str__(self):
+        degree_symbol = u'\N{DEGREE SIGN}'
+        return f"               x = {self.x:.3} m\n" \
+               f"               y = {self.y:.3} m\n" \
+               f"           Angle = {self.angle:.3}{degree_symbol}\n" \
+               f"        Velocity = {self.v:.3} m/s\n" \
+               f"Angular velocity = {math.degrees(self.w):.3} {degree_symbol}/s\n" \
+
+
+    @property
+    def x(self):
+        return self._x
+
+    @x.setter
+    def x(self, value):
+        self._x = value
+
+    @property
+    def y(self):
+        return self._y
+
+    @y.setter
+    def y(self, value):
+        self._y = value
+
+    @property
+    def angle(self):
+        return math.degrees(self._theta)
+
+    @angle.setter
+    def angle(self, value):
+        self._theta = math.radians(value)
+
+    @property
+    def theta(self):
+        return self._theta
+
+    @theta.setter
+    def theta(self, value):
+        self._theta = value
+
+    @property
+    def position(self):
+        return self.x, self.y
+
+    @position.setter
+    def position(self, value: tuple):
+        self.x, self.y = value
+
+    def reset_pose(self):
+        self.x, self.y, self.theta = (0.0, 0.0, 0.0)
+
+    def calc_distance(self, point_x, point_y):
+        dx = self.x - point_x
+        dy = self.y - point_y
+        return math.hypot(dx, dy)
+
+
+class GoalCriteria:
+    def __init__(self, full_speed_distance_error=0.02, full_speed_angle_error=4.0):
+        self._full_speed_distance_error = full_speed_distance_error
+        self._full_speed_angle_error = math.radians(full_speed_angle_error)
+
+        self._max_distance_error = None
+        self._max_angle_error = None
+
+    def angle(self, angle_error):
+        return abs(angle_error) < self._max_angle_error
+
+    def distance(self, distance_error, heading_error):
+        if abs(distance_error) < self._max_distance_error:
+            return True
+        if abs(heading_error) > math.pi / 2:
+            # Overshot goal, error will be small so consider goal to be reached.
+            # Otherwise robot would have to reverse a tiny amount which would be poor UX.
+            # Robot will correct itself on subsequent navigation calls anyway.
+            return True
+        return False
+
+    def update_linear_speed(self, speed_factor):
+        self._max_distance_error = speed_factor * self._full_speed_distance_error
+
+    def update_angular_speed(self, speed_factor):
+        self._max_angle_error = speed_factor * self._full_speed_angle_error
+
+
+class RobotDrivingManager:
```

Comment:
Name suggestions welcome, currently a bit ambiguous as to `drive_controller` vs `drive_manager`